### PR TITLE
[REEF-1672] Define Immutable Counters and Gauges

### PR DESCRIPTION
**Pull request by dkm2110**
This addressed the issue by
  * introducing Immutable gauges and counter classes.
  * introducing corresponding tests.

JIRA:
  [REEF-1672](https://issues.apache.org/jira/browse/REEF-1672)

### CODE REVIEW

**Comment by dkm2110** [comment]
@tcNickolas @markusweimer Please start reviewing this instead of REEF-1543.

File: lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        /// <summary>
+        /// Tests <see cref="ImmutableDoubleGauge"/>. Creates the class and verifies 
+        /// that all functions and properties work correctly.</summary>
+        [Fact]
+        public void TestImmutableDoubleGauge()
+        {
+            const string name = "default";
+            const string desc = "default";
+            const double value = 4;
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs]: Please use an actual double value to test a double gauge.

File: lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            var gauge = new ImmutableDoubleGauge(new MetricsInfoImpl(name, desc), value);
+            Assert.Equal(value, gauge.NumericValue);
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs]: Please don't use exact comparisons when comparing double values. This also applies to other places in code, including the code already checked-in in #1073 (`ImmutableMetricsImpl`). We already have tests failing due to exact comparisons of doubles, let's not repeat this.

**Comment by jwang98052** [comment]
@dkm2110 Do you have technical design document for Metrics? What each API is designed for? What they are supposed to do? What are the definition of Gauge, Tag, Rate, Stat, Metrics, Record, how they are related? What is the role of visitor? Why mutable/immutable are introduced here? I just have some difficulty to understand the code at the abstract level before it is associated to any REEF metrics. May be I missed some discussions before. A document or an overview with us would be welcome. Thanks!

File: lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/ImmutableCounter.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Org.Apache.REEF.Common.Metrics.Api;
+
+namespace Org.Apache.REEF.Common.Metrics.MetricsSystem
+{
+    /// <summary>
+    /// Immutable Counter implementation. This class is not thread safe and since it
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/ImmutableCounter.cs]: This is no any instance variables in this  class and no other method except constructor, what thread safe you mean here? 
If it is really thread unsafe, you might want to notate it as  [NotThreadSafe]

File: lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/ImmutableDoubleGauge.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Org.Apache.REEF.Common.Metrics.Api;
+
+namespace Org.Apache.REEF.Common.Metrics.MetricsSystem
+{
+    /// <summary>
+    /// Immutable double gauge implementation. This class is not thread safe and since it
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/ImmutableDoubleGauge.cs]: Can you put some description for what this class does?

File: lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs (hunk 1 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs]: done

File: lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs (hunk 2 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs]: done

**Comment by dkm2110** [comment]
@jwang98052 the functionalities are exactly based on subset of metrics2 design in hadoop.
You should look at that.
http://hadoop.apache.org/docs/r2.7.3/api/org/apache/hadoop/metrics2/package-summary.html

File: lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/ImmutableCounter.cs (hunk 3 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/ImmutableCounter.cs]: I see... let me remove thread safety issues from here.. it is actually not relevant here.

**Comment by dkm2110** [comment]
@tcNickolas @jwang98052 @markusweimer  I have addressed the review comments.

**Comment by dkm2110** [comment]
@tcNickolas  regarding failed appveyor build.. I had a look.....One Transport server related test failed that has nothing to do with this.

**Comment by jwang98052** [comment]
For metrics2 in hadoop, is there any Neget we can just use or we have to port the source code over? Is there any evaluation for why we choose this one?

**Comment by dkm2110** [comment]
@jwang98052  It's a java code.

**Comment by dkm2110** [comment]
@tcNickolas @jwang98052 Any further comments on this?

File: lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        /// <summary>
+        /// Tests <see cref="ImmutableLongGauge"/>. Creates the class and verifies 
+        /// that all functions and properties work correctly.
+        /// </summary>
+        [Fact]
+        public void TestImmutableLongGauge()
+        {
+            const string name = "default";
+            const string desc = "default";
+            const long value = 3;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs]: You should use a long value in the test otherwise it makes no difference from counter case.

File: lang/cs/Org.Apache.REEF.Common.Tests/metrics/MetricTestUtils.cs [hunk 6]
```diff
@@ -175,5 +175,32 @@ public IMetricsRecordBuilder CreateRecord(IMetricsInfo info)
...
+            public long LongGauge { get; private set; }
+
+            public double DoubleGauge { get; private set; }
+
+            public void Gauge(IMetricsInfo info, long value)
+            {
+                LongGauge = value;
+            }
+
+            public void Gauge(IMetricsInfo info, double value)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/MetricTestUtils.cs]: As this is the only implementation of IMetricsVisitor, you might want to show how IMetricsInfo is used in the method as an example.

**Comment by jwang98052** [comment]
I completed this pass with some minor comments.

File: lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs (hunk 5 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs]: @jwang98052  Counters are also using long. So I could not understand this comment......Main difference from counters is that we can only increment it while gauge can only decrement....Here we have to differentiate between long and gauge is only to make sure that for example later if filter needs to be applied to filter counters while long gauges can be passed through we will need that.

File: lang/cs/Org.Apache.REEF.Common.Tests/metrics/MetricTestUtils.cs (hunk 6 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/MetricTestUtils.cs]: I think this is orthogonal to the tests....Once we implement an actual/sample visitor we can do that.

**Comment by dkm2110** [comment]
@jwang98052 Responded to your comments. Please have a look.

File: lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs (hunk 5 above)
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs]: @dkm2110 Gauge can *only* decrement or gauge can *also* decrement? I though gauge can be used for keeping track of arbitrary values

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs]: "Note that there Counter has only Increment function while Gauges have both Increment and Decrement."
Both counter and gauge inherits ImmutableMetricsImpl and call base in constructor. Where can I find Increment and Decrement function for each?

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricImplTests.cs]: Please look at their mutable versions at: Org.Apache.REEF.Common/metrics/MutableMetricsLayer.
The immutable versions in this PR are not supposed to have these. Otherwise they will not be immutable.

**Comment by jwang98052** [comment]
LGTM

**Comment by dkm2110** [comment]
@jwang98052  @tcNickolas Can one of you merge?

**Comment by tcNickolas** [comment]
@dkm2110 Please resolve conflict this branch has before we merge

**Comment by tcNickolas** [comment]
I just noticed that this change also updates version of Visual Studio in `O.A.R.sln` file. Please revert this change. We'll update the version once we drop VS2013 support.

**Comment by dkm2110** [comment]
@tcNickolas I have excluded the .sln file change. Please merge if everything is fine.
Regarding REEF-1538, will send a separate one.
